src/protocol/ascii.rs

```diff
@@ -41,11 +41,98 @@ impl fmt::Display for StoreCommand {
         }
     }
 }
+
+struct CappedLineReader<C> {
+    inner: C,
+    filled: usize,
+    buf: Box<[u8]>,
+}
+
+fn get_line(buf: &[u8]) -> Option<usize> {
+    for (i, r) in buf.iter().enumerate() {
+        if *r == b'\r' {
+            if buf.get(i + 1) == Some(&b'\n') {
+                return Some(i + 2);
+            }
+        }
+    }
+    None
+}
+
+impl<C: Read> CappedLineReader<C> {
+    fn with_capacity(capacity: usize, inner: C) -> Self {
+        let buf = vec![0x0; capacity];
+        Self {
+            inner,
+            filled: 0,
+            buf: buf.into_boxed_slice(),
+        }
+    }
+
+    pub(crate) fn get_mut(&mut self) -> &mut C {
+        &mut self.inner
+    }
+
+    fn read_exact(&mut self, buf: &mut [u8]) -> Result<(), MemcacheError> {
+        let min = std::cmp::min(buf.len(), self.filled);
+        let (to_fill, rest) = buf.split_at_mut(min);
+        to_fill.copy_from_slice(&self.buf[..min]);
+        self.consume(min);
+        if rest.len() != 0 {
+            self.inner.read_exact(&mut rest[..])?;
+        }
+        Ok(())
+    }
+
+    /// Try to read a CRLF terminated line from the underlying reader.
+    /// The length of the line is expected to be <= the length of the
+    /// internal buffer, suited for reading headers or short responses.
+    ///
+    /// After reading a line, `consume` should be called.
```

Comment:
In hindsight, I think it's better to use a callback based approach, so that this doesn't need to be done all over the place. I'll change.